monashspa/PHS3302/d0_lifetime/d0_lifetime.py

```diff
@@ -0,0 +1,410 @@
+# 
+# 
+# The data sample we will use for this assignment was taken by the LHCb experiment using collisions between two beams of protons. We will only look at a very small fraction of the total sample. When the $D^0$ mesons are produced at the LHC they will be traveling relativistically. As a result of time-dilation the distance they typically travel before decaying can be as much as a few millimeters even though their life time is on the order of picoseconds. Most particles produced in proton-proton collisions are emitted directly from the point of the collision (primary vertex), therefore we can use the position of the $D^{0}$ decay (secondary vertex) to help disentagle them from the many other particles produced in the collision.
+# 
+# Notes:
+# - All momentums are given in units of ${\rm MeV}$ and distances in $\rm mm$.
+# 
+
+# ## Part 1: Reading in the data
+
+# To begin with we will download the data set and access the information we will need.
+# This file contains around 100,000 $D^0\rightarrow K^{-} \pi^{+}$ decays.
+# 
+# The file contains:
+# - the 3 components of momentum for the kaon and pion,
+# - the position of the proton-proton collision (primary vertex),
+# - the position of the $D^0$ decay (secondary vertex).
+
+
+
+import monashspa
+import uproot4 as uproot
+import matplotlib.pyplot as plt
+import numpy as np
+import math
+from monashspa.common.download_gdrive import download_file_from_google_drive
+
+## define constants
+const_c  = 2.99792e8 * 1e3 * 1e-12 ## in units mm/ps
+
+
+
+
+## Download the data
+download_file_from_google_drive('1sJI2nsJGmCu52xvpqCaijYQlDyqGA2b7','Data_small.root')
+
+## Access the information we need
+file = uproot.open("Data_small.root")
+tree = file["RS_LTUNB"]
+
+## momentum
+K_PX  = tree["P1_PX"].array(library="np")
+Pi_PX = tree["P2_PX"].array(library="np")
+
+K_PY  = tree["P1_PY"].array(library="np")
+Pi_PY = tree["P2_PY"].array(library="np")
+
+K_PZ  = tree["P1_PZ"].array(library="np")
+Pi_PZ = tree["P2_PZ"].array(library="np")
+
+## secondary vertex position
+D0_X = tree["D0_ENDVERTEX_X"].array(library="np")
+D0_Y = tree["D0_ENDVERTEX_Y"].array(library="np")
+D0_Z = tree["D0_ENDVERTEX_Z"].array(library="np")
+
+## primary vertex position
+PV_X = tree["D0_OWNPV_X"].array(library="np")
+PV_Y = tree["D0_OWNPV_Y"].array(library="np")
+PV_Z = tree["D0_OWNPV_Z"].array(library="np")
+
+
+
+
+## Lets try plotting the momentum of the kaon and pion
+fig, ax = plt.subplots(2, 3,figsize=(10,5))
+ax[0,0].hist(K_PX,bins=100,label="K_PX")
+ax[0,0].set_xlabel('$p_{x}(K^-)$ [MeV]')
+ax[0,0].set_ylabel('Candidates')
+
+ax[0,1].hist(K_PY,bins=100)
+ax[0,1].set_xlabel('$p_{y}(K^-)$ [MeV]')
+ax[0,1].set_ylabel('Candidates')
```

Comment:
I find px and py plots a bit hard to comprehend Would it be better to look at Pt?

Reply:
Good suggestion! That will also allow me to include an example of using numpy to calculate new quantities.